File: ios/Classes/agent/Mixer/RCFlutterAudioEffectManager.m
//
//  RCFlutterAudioEffectManager.m
//  rongcloud_rtc_plugin
//
//  Created by 潘铭达 on 2020/10/27.
//

#import "RCFlutterAudioEffectManager.h"
#import "RCFlutterAudioEffectManager+Apis.h"
#import "RCFlutterAudioEffectManager+Private.h"
#import "RCFlutterChannelKey.h"
#import "RCFlutterTools.h"
#import "RCFlutterEngine.h"

@interface RCFlutterAudioEffectManager () {
    NSMutableDictionary *_effects;
}

@property(nonatomic, strong) FlutterMethodChannel *channel;

@end

@implementation RCFlutterAudioEffectManager

SingleInstanceM(AudioEffectManager);

- (instancetype) init {
    self = [super init];
    if (self) {
        _effects = [NSMutableDictionary dictionary];
        [RCRTCEngine sharedInstance].audioEffectManager.delegate = self;
        [self registerChannel];
    }
    return self;
}

- (NSDictionary *)toDic {
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:[NSNumber numberWithInteger:0] forKey:@"id"];
    return dic;
}

- (void)destroy {
    [_effects removeAllObjects];
}

- (void)registerChannel {
    NSString *channelId = [NSString stringWithFormat:@"%@%d", KAudioEffectManager, 0];
    _channel = [FlutterMethodChannel methodChannelWithName:channelId
                                           binaryMessenger:[[RCFlutterEngine sharedEngine].pluginRegister messenger]];
    [[RCFlutterEngine sharedEngine].pluginRegister addMethodCallDelegate:self channel:_channel];
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([call.method isEqualToString:KPreloadEffect]) {
        [self preloadEffect:call result:result];
    } else if ([call.method isEqualToString:KUnloadEffect]) {
        [self unloadEffect:call result:result];
    } else if ([call.method isEqualToString:KPlayEffect]) {
        [self playEffect:call result:result];
    } else if ([call.method isEqualToString:KPauseEffect]) {
        [self pauseEffect:call result:result];
    } else if ([call.method isEqualToString:KPauseAllEffects]) {
        [self pauseAllEffects:result];
    } else if ([call.method isEqualToString:KResumeEffect]) {
        [self resumeEffect:call result:result];
    } else if ([call.method isEqualToString:KResumeAllEffects]) {
        [self resumeAllEffects:result];
    } else if ([call.method isEqualToString:KStopEffect]) {
        [self stopEffect:call result:result];
    } else if ([call.method isEqualToString:KStopAllEffects]) {
        [self stopAllEffects:result];
    } else if ([call.method isEqualToString:KSetEffectsVolume]) {
        [self setEffectsVolume:call result:result];
    } else if ([call.method isEqualToString:KGetEffectsVolume]) {
        [self getEffectsVolume:result];
    } else if ([call.method isEqualToString:KSetEffectVolume]) {
        [self setEffectVolume:call result:result];
    } else if ([call.method isEqualToString:KGetEffectVolume]) {
        [self getEffectVolume:call result:result];
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (void)preloadEffect:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSDictionary *arguments = (NSDictionary *)call.arguments;
    NSString *path = arguments[@"path"];
    NSString *assets = arguments[@"assets"];
    NSString *file = path != nil ? path : [self getAssetsPath:assets];
    NSInteger effectId = [arguments[@"effectId"] integerValue];
    [_effects setObject:file forKey:@(effectId)];
    RCRTCCode code = [self preloadEffect:effectId filePath:file];
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@(code) forKey:@"code"];
    NSString *json = [RCFlutterTools dictionaryToJson:dic];
    result(json);
    [self complete:code];
}

- (NSString *)getAssetsPath:(NSString *)assets {
    return [[NSBundle mainBundle] pathForResource:[[RCFlutterEngine sharedEngine].pluginRegister lookupKeyForAsset:assets] ofType:nil];
}

- (void)complete:(RCRTCCode)error {
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@(error) forKey:@"error"];
    NSString *json = [RCFlutterTools dictionaryToJson:dic];
    [self.channel invokeMethod:@"complete" arguments:json];
}

- (void)unloadEffect:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSDictionary *arguments = (NSDictionary *)call.arguments;
    NSInteger effectId = [arguments[@"effectId"] integerValue];
    [_effects removeObjectForKey:@(effectId)];
    RCRTCCode code = [self unloadEffect:effectId];
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@(code) forKey:@"code"];
    NSString *json = [RCFlutterTools dictionaryToJson:dic];
    result(json);
}

- (void)playEffect:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSDictionary *arguments = (NSDictionary *)call.arguments;
    NSInteger effectId = [arguments[@"effectId"] integerValue];
    NSString *path = [_effects objectForKey:@(effectId)];
    int loopCount = [arguments[@"loopCount"] intValue];
    NSUInteger volume = [arguments[@"volume"] unsignedIntegerValue];
    RCRTCCode code = [self playEffect:effectId filePath:path loopCount:loopCount publish:YES];
    [self setVolumeOfEffect:effectId withVolume:volume];
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@(code) forKey:@"code"];
    NSString *json = [RCFlutterTools dictionaryToJson:dic];
    result(json);
}

- (void)pauseEffect:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSDictionary *arguments = (NSDictionary *)call.arguments;
    NSInteger effectId = [arguments[@"effectId"] integerValue];
    RCRTCCode code = [self pauseEffect:effectId];
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@(code) forKey:@"code"];
    NSString *json = [RCFlutterTools dictionaryToJson:dic];
    result(json);
}

- (void)pauseAllEffects:(FlutterResult)result {
    RCRTCCode code = [self pauseAllEffects];
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@(code) forKey:@"code"];
    NSString *json = [RCFlutterTools dictionaryToJson:dic];
    result(json);
}

- (void)resumeEffect:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSDictionary *arguments = (NSDictionary *)call.arguments;
    NSInteger effectId = [arguments[@"effectId"] integerValue];
    RCRTCCode code = [self resumeEffect:effectId];
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@(code) forKey:@"code"];
    NSString *json = [RCFlutterTools dictionaryToJson:dic];
    result(json);
}

- (void)resumeAllEffects:(FlutterResult)result {
    RCRTCCode code = [self resumeAllEffects];
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@(code) forKey:@"code"];
    NSString *json = [RCFlutterTools dictionaryToJson:dic];
    result(json);
}

- (void)stopEffect:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSDictionary *arguments = (NSDictionary *)call.arguments;
    NSInteger effectId = [arguments[@"effectId"] integerValue];
    RCRTCCode code = [self stopEffect:effectId];
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@(code) forKey:@"code"];
    NSString *json = [RCFlutterTools dictionaryToJson:dic];
    result(json);
}

- (void)stopAllEffects:(FlutterResult)result {
    RCRTCCode code = [self stopAllEffects];
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@(code) forKey:@"code"];
    NSString *json = [RCFlutterTools dictionaryToJson:dic];
    result(json);
}

- (void)setEffectsVolume:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSDictionary *arguments = (NSDictionary *)call.arguments;
    NSUInteger volume = [arguments[@"volume"] unsignedIntegerValue];
    RCRTCCode code = [self setEffectsVolume:volume];
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@(code) forKey:@"code"];
    NSString *json = [RCFlutterTools dictionaryToJson:dic];
    result(json);
}

- (void)getEffectsVolume:(FlutterResult)result {
    NSUInteger volume = [self getEffectsVolume];
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@(volume) forKey:@"volume"];
    NSString *json = [RCFlutterTools dictionaryToJson:dic];
    result(json);
}

- (void)setEffectVolume:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSDictionary *arguments = (NSDictionary *)call.arguments;
    NSInteger effectId = [arguments[@"effectId"] integerValue];
    NSUInteger volume = [arguments[@"volume"] unsignedIntegerValue];
    RCRTCCode code = [self setVolumeOfEffect:effectId withVolume:volume];
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@(code) forKey:@"code"];
    NSString *json = [RCFlutterTools dictionaryToJson:dic];
    result(json);
}

- (void)getEffectVolume:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSDictionary *arguments = (NSDictionary *)call.arguments;
    NSInteger effectId = [arguments[@"effectId"] integerValue];
    NSUInteger volume = [self getVolumeOfEffectId:effectId];
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@(volume) forKey:@"volume"];
    NSString *json = [RCFlutterTools dictionaryToJson:dic];
    result(json);
}

- (void)didReportEffectPlayingProgress:(float)progress effectId:(NSUInteger)effectId {
    
}

- (void)didEffectFinished:(NSUInteger)effectId {
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@(effectId) forKey:@"effectId"];
    NSString *json = [RCFlutterTools dictionaryToJson:dic];
    [self.channel invokeMethod:@"onEffectFinished" arguments:json];
}

@end

File: android/src/main/java/io/rong/flutter/rtclib/agent/RCFlutterAudioEffectManager.java
package io.rong.flutter.rtclib.agent;

import android.util.Log;

import androidx.annotation.NonNull;

import com.alibaba.fastjson.JSONObject;

import cn.rongcloud.rtc.api.IAudioEffectManager;
import cn.rongcloud.rtc.api.IAudioEffectManager.IStateObserver;
import cn.rongcloud.rtc.api.IAudioEffectManager.ILoadingStateCallback;
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterAssets;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.rong.flutter.rtclib.utils.UIThreadHandler;

public class RCFlutterAudioEffectManager implements IStateObserver, ILoadingStateCallback, MethodCallHandler {

    public RCFlutterAudioEffectManager(BinaryMessenger messenger, FlutterAssets assets, IAudioEffectManager manager) {
        this.assets = assets;
        this.manager = manager;
        this.id = this.manager.hashCode();

        channel = new MethodChannel(messenger, "rong.flutter.rtclib/AudioEffectManager:" + this.id);
        channel.setMethodCallHandler(this);

        callback = this;
    }

    public int getId() {
        return id;
    }

    @Override
    public void onEffectFinished(int effectId) {
        JSONObject json = new JSONObject();
        json.put("effectId", effectId);
        String arguments = json.toString();
        UIThreadHandler.post(
                () -> channel.invokeMethod("onEffectFinished", arguments)
        );
    }

    @Override
    public void complete(int error) {
        JSONObject json = new JSONObject();
        json.put("error", error);
        String arguments = json.toString();
        UIThreadHandler.post(
                () -> channel.invokeMethod("complete", arguments)
        );
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        switch (call.method) {
            case "preloadEffect":
                preloadEffect(call, result);
                break;
            case "unloadEffect":
                unloadEffect(call, result);
                break;
            case "playEffect":
                playEffect(call, result);
                break;
            case "pauseEffect":
                pauseEffect(call, result);
                break;
            case "pauseAllEffects":
                pauseAllEffects(result);
                break;
            case "resumeEffect":
                resumeEffect(call, result);
                break;
            case "resumeAllEffects":
                resumeAllEffects(result);
                break;
            case "stopEffect":
                stopEffect(call, result);
                break;
            case "stopAllEffects":
                stopAllEffects(result);
                break;
            case "setEffectsVolume":
                setEffectsVolume(call, result);
                break;
            case "getEffectsVolume":
                getEffectsVolume(result);
                break;
            case "setEffectVolume":
                setEffectVolume(call, result);
                break;
            case "getEffectVolume":
                getEffectVolume(call, result);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    private void preloadEffect(MethodCall call, Result result) {
        String path = call.argument("path");
        String assets = call.argument("assets");
        String file = path != null ? path : getAssetsPath(assets);
        Integer effectId = call.argument("effectId");
        assert effectId != null : "Effect id can't be null!";
        manager.preloadEffect(file, effectId, error -> {
            JSONObject json = new JSONObject();
            json.put("code", error);
            UIThreadHandler.success(result, json.toJSONString());
        });
    }

    private String getAssetsPath(String assets) {
        return "file:///android_asset/" + this.assets.getAssetFilePathByName(assets);
    }

    private void unloadEffect(MethodCall call, Result result) {
        Integer effectId = call.argument("effectId");
        assert effectId != null : "Effect id can't be null!";
        int code = manager.unloadEffect(effectId);
        JSONObject json = new JSONObject();
        json.put("code", code);
        UIThreadHandler.success(result, json.toJSONString());
    }

    private void playEffect(MethodCall call, Result result) {
        Integer effectId = call.argument("effectId");
        assert effectId != null : "Effect id can't be null!";
        Integer loopCount = call.argument("loopCount");
        assert loopCount != null : "LoopCount can't be null!";
        Integer volume = call.argument("volume");
        assert volume != null : "Volume can't be null!";
        int code = manager.playEffect(effectId, loopCount, volume);
        JSONObject json = new JSONObject();
        json.put("code", code);
        UIThreadHandler.success(result, json.toJSONString());
    }

    private void pauseEffect(MethodCall call, Result result) {
        Integer effectId = call.argument("effectId");
        assert effectId != null : "Effect id can't be null!";
        int code = manager.pauseEffect(effectId);
        JSONObject json = new JSONObject();
        json.put("code", code);
        UIThreadHandler.success(result, json.toJSONString());
    }

    private void pauseAllEffects(Result result) {
        int code = manager.pauseAllEffects();
        JSONObject json = new JSONObject();
        json.put("code", code);
        UIThreadHandler.success(result, json.toJSONString());
    }

    private void resumeEffect(MethodCall call, Result result) {
        Integer effectId = call.argument("effectId");
        assert effectId != null : "Effect id can't be null!";
        int code = manager.resumeEffect(effectId);
        JSONObject json = new JSONObject();
        json.put("code", code);
        UIThreadHandler.success(result, json.toJSONString());
    }

    private void resumeAllEffects(Result result) {
        int code = manager.resumeAllEffects();
        JSONObject json = new JSONObject();
        json.put("code", code);
        UIThreadHandler.success(result, json.toJSONString());
    }

    private void stopEffect(MethodCall call, Result result) {
        Integer effectId = call.argument("effectId");
        assert effectId != null : "Effect id can't be null!";
        int code = manager.stopEffect(effectId);
        JSONObject json = new JSONObject();
        json.put("code", code);
        UIThreadHandler.success(result, json.toJSONString());
    }

    private void stopAllEffects(Result result) {
        int code = manager.stopAllEffects();
        JSONObject json = new JSONObject();
        json.put("code", code);
        UIThreadHandler.success(result, json.toJSONString());
    }

    private void setEffectsVolume(MethodCall call, Result result) {
        Integer volume = call.argument("volume");
        assert volume != null : "Volume can't be null!";
        int code = manager.setEffectsVolume(volume);
        JSONObject json = new JSONObject();
        json.put("code", code);
        UIThreadHandler.success(result, json.toJSONString());
    }

    private void getEffectsVolume(Result result) {
        int volume = manager.getEffectsVolume();
        JSONObject json = new JSONObject();
        json.put("volume", volume);
        UIThreadHandler.success(result, json.toJSONString());
    }

    private void setEffectVolume(MethodCall call, Result result) {
        Integer effectId = call.argument("effectId");
        assert effectId != null : "Effect id can't be null!";
        Integer volume = call.argument("volume");
        assert volume != null : "Volume can't be null!";
        int code = manager.setEffectVolume(effectId, volume);
        JSONObject json = new JSONObject();
        json.put("code", code);
        UIThreadHandler.success(result, json.toJSONString());
    }

    private void getEffectVolume(MethodCall call, Result result) {
        Integer effectId = call.argument("effectId");
        assert effectId != null : "Effect id can't be null!";
        int volume = manager.getEffectVolume(effectId);
        JSONObject json = new JSONObject();
        json.put("volume", volume);
        UIThreadHandler.success(result, json.toJSONString());
    }

    private final FlutterAssets assets;

    private final IAudioEffectManager manager;
    private final int id;

    private final MethodChannel channel;

    private final ILoadingStateCallback callback;
}
